Added: Pull request for "fsensor" and "contrtrol" dedicated control theory OOP

dvdvideo1234 opened the pull request:
Added: "fsensor" as a general environment sensor for obtaining process data
Added: "contrtrol" for creating a OOP process variable controller that can me tuned using various methods

Based on: https://github.com/wiremod/wire-extras/pull/59

dvdvideo1234 commented:
@thegrb93 FYI

Tell me if you need some adjustments. The addon is updated to the latest version from my personal machine. It includes the all recent modifications. You can test it if you like :+1:

thegrb93 commented:
Firstly get rid of all the concatenation in the e2 description file unless there's a really good reason for it.

dvdvideo1234 commented:
The idea is if something is gonna get changed in the future, it will be an update of only one value.
You can read them here: https://github.com/dvdvideo1234/ControlSystemsE2/wiki

This wiki is automatically generated from the description ;)

dvdvideo1234 commented:
Exported and replaced to ease you out
[fsensor_descr.txt](https://github.com/wiremod/wire-extras/files/3457211/fsensor_descr.txt)
[stcontrol_descr.txt](https://github.com/wiremod/wire-extras/files/3457212/stcontrol_descr.txt)

thegrb93 commented:
I'd say if you really needed to change anything, a find-replace-all would be just as easy. I'll look at your wiki page for now. My main concern is code bloat i.e. adding functions nobody will ever use and just get in the way, or functions that are just as easy to make inside e2 itself. I'll look tonight.

Divran commented:
@thegrb93 keep in mind that this is wire extras, where we don't need to care as much about bloat and such. Still care a little, just not as much as the normal wire repo

dvdvideo1234 commented:
Done. Just tell me how you like 'em ok ;)

thegrb93 commented:
That looks better, flash sensor needs a new name though. I have coded and played gmod for 12 years and I had no idea what it was until I read the wiki page. It's Trace structure basically, right? Just call it a Trace structure.

thegrb93 commented:
Don't do manual collectgarbage()'s

dvdvideo1234 commented:
Gotta leave them to ghe game  collector then.

dvdvideo1234 commented:
Removing the `collectgarbage` calls can be done only by removing the flag from `remValue` so I will do just that. Thanks !

dvdvideo1234 commented:
> That looks better, flash sensor needs a new name though. I had no idea what it was until I read the wiki page. It's Trace structure basically, right? Just call it a Trace structure.

Basically, an OOP wich traces and retrieves trace result. Later you can read every aspect of this trace result. Kinda like ranger data, but not need to calculate positions directions, filters and all that stuff in every to create a `RD` class. My tracer is initialized only once and uses this stuff internally ;)

thegrb93 commented:
Get rid of outError and outPrint or make them singleplayer only.

dvdvideo1234 commented:
@thegrb93 

What I should replace them with ?  print() / error() ?

thegrb93 commented:
Either nothing or the core print function.

dvdvideo1234 commented:
How can I output an error then ?

Divran commented:
All E2 functions fail silently and/or return a zero equivalent value whenever there's an error. If you call the lua error function you shut down the entire chip, which is something that E2 tries to avoid.
Other E2 functions that can error, and where the error message is important (such as json decode or whatever) always have a "get error" function that you can check after.
There's very very few things in E2 that break this rule, and arguably some of them could also be changed to follow it.

The print function seems fine, but pointless. You're using the standard lua print function, and since E2 runs serverside, the user of the E2 will never see those prints. So for example the `dumpConsole` function literally does nothing.

dvdvideo1234 commented:
Can I Leave the print statement then, to be intended for singleplayer? 
Could you please tell me what I need to call to print in the console?

thegrb93 commented:
Look in the debug.lua for how e2 printing works https://github.com/wiremod/wire/blob/master/lua/entities/gmod_wire_expression2/core/debug.lua

bigdogmat commented:
> Or I just return `0` or `nil` when something fails and that's it?

Depends on the type, it's whatever you think a good default is for the value. e.g. number = 0, string = "", entity = NULL

dvdvideo1234 commented:
Thank you @thegrb93 @bigdogmat. I have to test how the different prints behave in my home machine because I want to be able to tell the user that something happened.

dvdvideo1234 commented:
Ok. [I was able to replace every occurrence of `fsensor` with `ftracer`](https://github.com/dvdvideo1234/ControlSystemsE2/commit/c384134f5aa90ccd8e8efcff1c652f75cc406053) because seems more correct that way. The OOP is basically an object that does internal tracing and gathers environment data.

dvdvideo1234 commented:
Which of these will be better to replace the printing functions:
1. [Wire's Addnotify](https://github.com/wiremod/wire/blob/master/lua/entities/gmod_wire_expression2/core/debug.lua#L183) - The addon will sow messages in the notification area, but the chip will not halt
2. [Player char print](https://github.com/wiremod/wire/blob/master/lua/entities/gmod_wire_expression2/core/debug.lua#L137) - Output errors in the chat area, which may be skipped by the user if the server relies on chat communication.
3. [Player print message](https://github.com/wiremod/wire/blob/master/lua/entities/gmod_wire_expression2/core/debug.lua#L202) - This seems to be a great alternative to (2.) but with more output options.

So far I'd like to use option (1.), but if the chip execution does not stop will be more likely has to be (3.), because the screen will be flooded with hint messages otherwise. What do you guys think ?

thegrb93 commented:
Chatprint probably.

dvdvideo1234 commented:
@thegrb93 
Hello, the lasted version has been merged. I removed `outPrint` and `outError` as you sugested and replaced them with `logStatus` which just prints the error messages to various outputs ( def. CHAT ) without halting the E2.

thegrb93 commented on lua/entities/gmod_wire_expression2/core/custom/stcontrol.lua:
```diff
@@ -0,0 +1,1094 @@
+end
+
+local function isString(vS)
+  return (getmetatable(vS) == gtStringMT)
```
isstring function already exists. Use that instead as it's much more efficient than this.

thegrb93 commented:
You still didn't remove the collectgarbages

dvdvideo1234 commented:
@thegrb93 Done. I removed also the metatables as they are not used.

dvdvideo1234 commented:
Sorry, I cannot test it now because I am on the phone now. I tested it yesterday in SP.

dvdvideo1234 commented:
@thegrb93 Thought I removed them hmm.. Just found two in the controller OOP. Guessential I removed them from the tracer. Now I removed all of them.

thegrb93 commented:
Ok I'll take another look in a bit

thegrb93 commented on lua/entities/gmod_wire_expression2/core/custom/stcontrol.lua:
```diff
@@ -0,0 +1,1088 @@
+local gtMissName  = {"Xx", "X", "Nr"} -- This is a place holder for missing/default type
+local gtStoreOOP  = {} -- Store state controllers here linked to the entity of the E2
+local gsVarPrefx  = "wire_expression2_stcontrol" -- This is used for variable prefix
+local varMaxTotal = CreateConVar(gsVarPrefx.."_max" , 20, gnServerControled, "StControl items maximum count")
```
Why is there a max? That seems unnecessary.

dvdvideo1234 replied:
Just like max wire rangers. Same logic for controllers too. I want them to be somewhat identical.

thegrb93 replied:
For what reason does there need to be a max though? Just let quota limit them.

dvdvideo1234 replied:
Hello,

Sorry that I write so late. I literally had no time to spend on this in the close weeks.

Do you mean this `Tick quota exceeded`, i think this stops the E2 and you tried to avoid such events ... I am kinda worried about the extension might hang the server if many people create many controllers/tracers and start to request calculation/sampling. I already put the `setState` function to have  `__e2setcost(20)` for example. Do you have something else in mind and if so how should code it ?

Thanks !

thegrb93 commented on lua/entities/gmod_wire_expression2/core/custom/stcontrol.lua:
```diff
@@ -0,0 +1,1088 @@
+  tSrc[aKey] = nil; return tSrc
+end
+
+local function logStatus(sMsg, oSelf, nPos, ...)
```
This needs to be removed or limited. It can potentially lag the server or disconnect the player

dvdvideo1234 replied:
Yes. Right now its limited if the user have it enabled client side.

thegrb93 replied:
Looks like its a serverside convar actually. I guess that's ok.

dvdvideo1234 replied:
This can be easily disabled ( by default being disabled ) via convar `varEnStatus` reading `wire_expression2_stcontrol_enst` and since there is one on the server and also client not replicating each other on MP player can run it on the client so the server does not bother at all with that ;)

thegrb93 commented on lua/entities/gmod_wire_expression2/core/custom/stcontrol.lua:
```diff
@@ -0,0 +1,1088 @@
+      gtPrintName["TALK"   ] = HUD_PRINTTALK
+      gtPrintName["CENTER" ] = HUD_PRINTCENTER
+
+local function isEntity(vE)
```
Should be renamed to isValidEntity or just isValid

dvdvideo1234 replied:
OK.

dvdvideo1234 replied:
I am OK with isValidEntity

dvdvideo1234 replied:
Changed to `isValid` local function reference

thegrb93 commented on lua/entities/gmod_wire_expression2/core/custom/stcontrol.lua:
```diff
@@ -0,0 +1,1088 @@
+  return (vE and vE:IsValid())
+end
+
+local function isHere(vV)
```
Useless helper function. Remove it; makes code less readable.

dvdvideo1234 replied:
It is helper function. I need to check for Nil. This is because if some parameters ate present they will be used and they are ignored of they are nil. This is for not putting extra flags for use/don't use

dvdvideo1234 replied:
Removed. Using `if(V)` and `if(not V)`.

thegrb93 commented on lua/entities/gmod_wire_expression2/core/custom/stcontrol.lua:
```diff
@@ -0,0 +1,1088 @@
+  return (kV*getSign(eV)*mathAbs(eV)^pV)
+end
+
+local function remValue(tSrc, aKey)
```
Useless helper function. Remove it; makes code less readable.

dvdvideo1234 replied:
Remove value? I use it to assign nil. Though everywhere I assign I prefer a function for it. Make the code more facorized.

thegrb93 replied:
It forces whoever reads the code to learn what the function does and memorize it when all they should have to see is x.y = nil. Most helper functions are good, but this one is unnecessary.

dvdvideo1234 replied:
Ok. Seems fair because it mes it `NIL` anyways ;)

dvdvideo1234 replied:
Changed, thanks !

thegrb93 commented on lua/entities/gmod_wire_expression2/core/custom/stcontrol.lua:
```diff
@@ -0,0 +1,1088 @@
+  end; return ...
+end
+
+local function getControllersCount() local mC = 0
```
I don't think we need to be keeping track of how many there are.

dvdvideo1234 replied:
Does this mean that the convar for maximum amount is not needed.

thegrb93 replied:
Yes

dvdvideo1234 replied:
Does this mean that I can again use `quota limit` for this as for the tracers ?

thegrb93 replied:
I don't know what that is. I was thinking just the E2 ops limit.

dvdvideo1234 replied:
I am afraid that too many [controlles](https://en.wikipedia.org/wiki/Control_loop) may lag the server. How can I use the `E2 ops limit` ?

dvdvideo1234 commented:
@thegrb93 

Hello, updated the helper functions according to your expectations. I will fix my separate addon once this PR is merged ;)

thegrb93 commented:
I'll take another look now that it's been updated.

dvdvideo1234 commented:
@thegrb93 

> 
> I'll take another look now that it's been updated.

Please tell me how to replace the items count with OPS and it will become near perfect :)

thegrb93 commented:
If you set the ops usage of the functions appropriately, it shouldn't matter how many of them the player makes, right?

dvdvideo1234 commented:
Is this actually the `__e2setcost(number nCost)` function ?

thegrb93 commented:
whatever keeps the server from lagging.

thegrb93 commented:
What is the purpose of gtStoreOOP? I'll look for myself later as well.

dvdvideo1234 commented:
It is for storing the OOP and keeps track which objects are created from which E2, so when the E2 is removed, the objects be removed with it. Also, you can call OOP:semSelf() to destroy the object.

thegrb93 commented:
That doesn't make any sense. Any memory inside the e2 gets removed with it anyway.

dvdvideo1234 commented:
At the moment I am using `gtStoreOOP` to store the controllers on an `E2` in the format:

```lua
gtStoreOOP = {}
gtStoreOOP[eChip][1] = E2:ControllerOOP_1 -- ( controller objects )
gtStoreOOP[eChip][2] = E2:ControllerOOP_2
-- All the way up to ...
gtStoreOOP[eChip][N] = E2:ControllerOOP_N
```

When the chip is removed, we have `eChip:CallOnRemove("stcontrol_remove_ent", remControllersEntity)` which executes the function:

```lua
local function remControllersEntity(eChip)
  if(not isValid(eChip)) then return end
  local tCon = gtStoreOOP[eChip]; if(not next(tCon)) then return end
  local mCon = #tCon; for ID = 1, mCon do tableRemove(tCon) end
end
```

Are you suggesting that I should store the controller/trqacer created objects inside some hash of the `E2` chip `oSelf`. For example the entity of the `E2` is `local eChip = oSelf.entity` and I must create somethin similar related to the addon. For example being the unique class name of the OOP:

```lua
local oStCon = {} -- The self table of OOP
local tStore = oSelf[eChip]["stcontrol"] = {}
tStore[#tStore+1] = oStCon
```

thegrb93 commented:
Would it not be possible to have them stored like other E2 classes? E.g. the default E2 ranger class.

dvdvideo1234 commented:
So you talk about:
```lua
local function ranger(self, rangertype, range, p1, p2, hulltype, mins, maxs, traceEntity )
	local data = self.data
	local chip = self.entity

	local defaultzero = data.rangerdefaultzero
	local ignoreworld = data.rangerignoreworld
	local water = data.rangerwater
	local entities = data.rangerentities
	local filter = data.rangerfilter
```
I have to loacalize this general `data` table and store my stuff in it i suppose `local data = self.data`.

thegrb93 commented:
No, look.
```lua
e2function ranger ranger(distance)
	return ranger(self, 0, distance, 0, 0) -- type 0, no skew
end
```
All it does is return a table. The 'ranger' in the e2function tells e2 what type the table is, and the table contains all of the ranger data.

dvdvideo1234 commented:
Done. Thank you !

thegrb93 commented on lua/entities/gmod_wire_expression2/core/custom/ftracer.lua:
```diff
@@ -0,0 +1,887 @@
+
+__e2setcost(1)
+e2function ftracer noFTracer()
+  return nil
```
Should this return an empty table instead of nil?

thegrb93 replied:
Nevermind it's fine.

thegrb93 commented:
Ok, this is looking a lot better. I can do a skim through it again and see if anything else looks off, but seems ready from a glance.

dvdvideo1234 commented:
Thank you very much. 😍

thegrb93 commented:
I'm not seeing any of the ftracer functions. Do I have to enable it or something?

thegrb93 commented:
Ah

[ERROR] entities/gmod_wire_expression2/core/custom/ftracer.lua:180: '=' expected near ';'
  1. unknown - addons/wire/lua/entities/gmod_wire_expression2/core/extloader.lua:86
   2. pcall - [C]:-1
    3. e2_include_finalize - addons/wire/lua/entities/gmod_wire_expression2/core/extloader.lua:101
     4. unknown - addons/wire/lua/entities/gmod_wire_expression2/core/extloader.lua:172
      5. include - [C]:-1
       6. wire_expression2_reload - addons/wire/lua/entities/gmod_wire_expression2/core/extloader.lua:37
        7. unknown - addons/wire/lua/entities/gmod_wire_expression2/core/e2lib.lua:583

dvdvideo1234 commented:
Let me check. I did not do Gmod in the past week.

dvdvideo1234 commented:
@thegrb93  Thank you! The spacing is conveted to tab also to match the code base !

thegrb93 commented:
Ok. I'll look again

dvdvideo1234 commented:
I also updated the line endings to CR-LF ( Windows ) to match the other files

dvdvideo1234 commented:
Updated the wiki accordingly: https://github.com/dvdvideo1234/ControlSystemsE2/wiki/FTrace

dvdvideo1234 commented:
@thegrb93 I will be able to tes it today and will show you the example code. Thanks !

thegrb93 commented:
Ok

dvdvideo1234 commented:
@thegrb93  Tested and updated with 2 bug fixes. You can find the sample code here:
https://github.com/dvdvideo1234/ControlSystemsE2/tree/master/data/Expression2

dvdvideo1234 commented:
Hello, @thegrb93, the source is ready to be merged on my side ;)
